### PR TITLE
compiler: Fix tensor_from_list method of DenseIntOrFPElementsAttr to properly handle f32 input values

### LINKED ISSUE #179: Fix DenseIntOrFPElementsAttr constructor
If given integer values, but a float type, to `tensor_from_list` of `DenseIntOrFPElementsAttr`, the values won't be converted to float values.

### PR DESCRIPTION
**Pull request by meshtag**
Fixes #179 

The following program can be used to verify that the proposed changes work as expected : 
```
from xdsl.dialects.builtin import *

def test_DenseIntOrFPElementsAttr():
    check1 = DenseIntOrFPElementsAttr.tensor_from_list([4, 5], f32)

    value1 = check1.data.data[0].value.data
    value2 = check1.data.data[1].value.data

    # Ensure type conversion happened properly during attribute construction.
    res1 = type(value1) == float and value1 == 4.0 and type(
        value2) == float and value2 == 5.0
    assert (res1)

    t1 = FloatAttr.from_value(4.0, f32)
    t2 = FloatAttr.from_value(5.0, f32)

    check2 = DenseIntOrFPElementsAttr.tensor_from_list([t1, t2], f32)

    value3 = check2.data.data[0].value.data
    value4 = check2.data.data[1].value.data

    # Ensure type conversion happened properly during attribute construction.
    res2 = type(value3) == float and value3 == 4.0 and type(
        value4) == float and value4 == 5.0
    assert (res2)


if __name__ == '__main__':
    test_DenseIntOrFPElementsAttr()
```

### CODE REVIEW

**Comment by webmiche** [comment]
Could you add the suggested test program as a test? There exists a `tests` directory where you can find other test files. Just add it somewhere you think it belongs or introduce a new one :)

**Comment by meshtag** [comment]
Actually, I asked whether we should add unit tests for python constructs of existing dialects in `xdsl` in the Zulip workspace chat of the project. @math-fehr responded by explaining that the current goal is to generate `MLIR` core dialects directly from TableGen, the existing python infrastructure might be tweaked in the future for this purpose and hence we should start working on these unit tests once the construct design is finalised. 

Nevertheless, I have no objection to adding a test for the above change and will add it if required.

**Review by math-fehr — APPROVED** [review]
Looks good!

**Comment by webmiche** [comment]
IMO, a bugfix should always be accompanied by a regression test case. If we actually move to autogenerating the dialects, we can still delete the tests for the current dialects in the same PR. So, IMO, the best way to go is creating a dialects directory in the tests directory and move this test there. But I am fine with following @math-fehr s lead here.

**Review by georgebisbas — CHANGES_REQUESTED** [review]
I am strongly in favour of adding the test...and removing it in the future if we consider so.
Otherwise, we may do changes that will break this and have no clue.

**Comment by meshtag** [comment]
I have added the test as discussed above. 

Shall I create a feature request for adding tests related to all `xdsl` dialects? 
This would help us keep a track of covered dialects and would also let everyone know which dialects are free to chose from (for adding tests).

File: tests/dialects/builtin_test.py
Comment (webmiche) [inline tests/dialects/builtin_test.py:13]: AFAIU, you are not using `res1` anywhere else. Therefore, I think it should be replaced by its definition. Additionally, I think it is easier if each individual part of this four-way conjunction is its own assertion. Then one immediately sees which condition was wrong.

Comment (webmiche) [reply tests/dialects/builtin_test.py:13]: Same for res2 and the other assertion.

File: tests/dialects/builtin_test.py
Comment (webmiche) [inline tests/dialects/builtin_test.py:4]: Maybe name this something like `test_DenseIntOrFPElementsAttr_fp_type_conversion` as you are testing a specific part of the DenseIntOrFPElementsAttr.

File: tests/dialects/builtin_test.py
Comment (webmiche) [inline tests/dialects/builtin_test.py:29]: These two lines are not needed in the tests. Pytest drives them already.

File: tests/dialects/builtin_test.py
Comment (webmiche) [inline tests/dialects/builtin_test.py:1]: I know this is not yet the case in all of the repo, but *-imports are generally considered bad style as there can be shadowing and other issues. Please import only what you need.

**Review by georgebisbas — CHANGES_REQUESTED** [review]
Could you rename the commits according to:
https://github.com/xdslproject/xdsl/wiki/Tags-for-commit-messages-and-PR-titles ?

**Comment by georgebisbas** [comment]
I understand the worries about the effort.
From this perspective one may not commit to this repo whatever may be dropped in the future.
A good way of CI/CD requires tests for this repo, irrespective of what is tested on other repos.
Assuming I am building on xdsl, "(I should not)" ""care"" of what is tested in ChocopY.
I hope you get my point here.

File: tests/dialects/builtin_test.py
Comment (webmiche) [inline tests/dialects/builtin_test.py:11]: NIT: I think you can remove the outermost brackets here.

**Review by webmiche — APPROVED** [review]

**Comment by meshtag** [comment]
> Could you rename the commits according to:
> https://github.com/xdslproject/xdsl/wiki/Tags-for-commit-messages-and-PR-titles ?

@georgebisbas , I renamed the latest commit (assuming all other commits will be squashed), can you please confirm if it works.

File: xdsl/dialects/builtin.py
Comment (math-fehr) [inline xdsl/dialects/builtin.py:580]:
```suggestion
            data: List[int | float | AnyFloatAttr]) -> DenseIntOrFPElementsAttr:
```
since we now also accept ints

**Review by math-fehr — APPROVED** [review]
Looks good!
I just added a small typing comment

File: xdsl/dialects/builtin.py
Comment (webmiche) [reply xdsl/dialects/builtin.py:580]: Shouldn't we then also accept `AnyIntegerAttr` to be consistent?

Comment (meshtag) [reply xdsl/dialects/builtin.py:580]: In that case, does it makes sense to change the name of the function as well? (Its current name is `from_float_list`)

Also, I might be a bit slow to respond in coming weeks since I am having my end-of-semester examinations at my university.

**Comment by webmiche** [comment]
> Hi again, PR looks good, but commits are not renamed properly. Try to follow this: https://github.com/xdslproject/xdsl/wiki/Tags-for-commit-messages-and-PR-titles
> 
> e.g.:
> 
> Add tests --> tests: Add tests

Considering that we are planning to use `squash and merge` from now on, only the title of the PR matters, as it becomes the commit message, right?

File: xdsl/dialects/builtin.py
Comment (math-fehr) [reply xdsl/dialects/builtin.py:580]: We can rename it `create_dense_float` instead

Comment (math-fehr) [reply xdsl/dialects/builtin.py:580]: We should also change the `from_list` type to also allow `int` and `AnyIntegerAttr`

Comment (meshtag) [reply xdsl/dialects/builtin.py:580]: > We can rename it create_dense_float instead,

Shall we also rename `from_int_list` to `create_dense_int` for uniformity then?

Comment (math-fehr) [reply xdsl/dialects/builtin.py:580]: Yes!

**Review by webmiche — APPROVED** [review]
LGTM! I guess the new constructors will have to be tested as well, but this can happen in a different PR, IMO.

File: xdsl/dialects/builtin.py
Comment (webmiche) [inline xdsl/dialects/builtin.py:592]: Actually, thinking about it again, shouldn't this also allow int and AnyIntegerAttr? This is the case that gets forwarded to `create_dense`, so we explicitely want to allow mixed lists, right?

File: xdsl/dialects/builtin.py
Comment (math-fehr) [inline xdsl/dialects/builtin.py:580]: Actually, `AnyIntegerAttr` should not be accepted here, having one value of this type would trigger an exception (trying to translate an `IntegerAttr` to a `float`).

File: xdsl/dialects/builtin.py
Comment (math-fehr) [reply xdsl/dialects/builtin.py:592]: Yes, this should also add `int`.

**Review by math-fehr — APPROVED** [review]
Looks good, there is just 1-2 problems with the types, but then we can merge it!

File: xdsl/dialects/builtin.py
Comment (webmiche) [reply xdsl/dialects/builtin.py:580]: True, though I feel that it should be if we were to allow ìnt` here...

Comment (math-fehr) [reply xdsl/dialects/builtin.py:580]: I am not completely sure if we would like to do that.
I feel here, it makes sense to go from `int` to `FloatAttr` (given a type), since we are converting an arbitrary-sized integer to a float. However, a conversion from `IntegerAttr` to `FloatAttr` makes a little bit less sense in my opinion, since we are now converting something that is bounded (and might have been converted already).
So maybe it would be dangerous to do this conversion, but maybe I'm also wrong and I'm overthinking this.

Comment (math-fehr) [reply xdsl/dialects/builtin.py:580]: So I don't have a strong opinion, either we add the conversion, or we change the type hint.

Comment (webmiche) [reply xdsl/dialects/builtin.py:580]: I am fine with the latter and potentially addressing this issue once it actually arises somewhere...

**Review by math-fehr — APPROVED** [review]
Nice!
That looks good to me now!

**Comment by math-fehr** [comment]
@georgebisbas, could you look if this is good to you? Then we can squash and merge

**Comment by webmiche** [comment]
I think this needs a rebase first, right?

**Review by georgebisbas — COMMENTED** [review]
@math-fehr PRs like this expose our constructors not being homogenous.
We have builders with from_ create_... it is not user friendly...thoughts?

**Review by georgebisbas — COMMENTED** [review]
Try to rebase first, yes, and then we can have a final review

**Comment by math-fehr** [comment]
So the question is, what naming scheme should we follow?
We could all start with `from`, but this makes only sense when we can describe a constructor from its arguments.
Here however, we want to describe the constructor with what it is constructing, unless you have a better naming starting with `from` ?

**Comment by georgebisbas** [comment]
> So the question is, what naming scheme should we follow? We could all start with `from`, but this makes only sense when we can describe a constructor from its arguments. Here however, we want to describe the constructor with what it is constructing, unless you have a better naming starting with `from` ?

Yes, this may need some thought, also possibly towards release v1.0...
Either way this needs to be properly documented...how can a user know how to construct something...
Maybe we should also create a discussion about this...not sure..

**Comment by webmiche** [comment]
I feel the problem really lies in the direction you look into. Sometimes, we want to describe the argument, sometimes the result is more central. I feel that we won't be able to unify this in the whole repo, unless we have both in each constructor, which gives extremely long constructors. Obviously, the notation `Class(Args)` sidesteps this issue by not giving a name to a constructor...

Let's take this discussion to (a) a github discussion or (b) the January meeting. I feel this PR has been open for long enough and should be merged soon. IMO, we can just keep the current state and iterate on it after the discussion...

**Comment by math-fehr** [comment]
I would guess that for that, we would need 1) to generate a documentation based on the code and 2) document what kind of constructors are expected for the users.
In either way, @georgebisbas, if you think that we can merge it, could you approve it? (otherwise we can't merge it)

**Review by georgebisbas — APPROVED** [review]
Approve

**Pull request merged**
